### ISSUE TITLE
Debug build exit code 3, `FIXME dDebug NEEDS UPDATING`

### ISSUE THREAD

**Issue by rahulraman0108** [efroemling/ballistica#306]
## Description
Debug builds quit with exit code 3 and non-descriptive and confusing message in output. `FIXME dDebug NEEDS UPDATING`. However, Release builds work fine. I tested with both GUI and server builds.
I was trying to update the portal actor script to the latest API and test it with the latest ballistica build, I have attached the code below and screenshots of output. The Debug builds error as soon as a Portal object is created however Release builds work fine.
The make `preflight` target runs complete and returns `PREFLIGHT SUCCESSFUL!`.

## Steps to reproduce
1. Add the attached script to the ballistica python scripts directory.
2. Add the code to create portal actor somewhere in root scripts (eg. a timer call to instantiate portal object in activity on_begin)
3. Build and launch a debug version of ballisticacore
4. Do defined actions to create the portal object
5. Bug!

## Expected behavior
If there is any error the console should have shown a descriptive error report otherwise, it should not crash the build.

## Machine
**Platform**: Windows 10 with WSL Ubuntu 20.04 LTS version 1
**Commit**: [63a835a1f4d4901481a4391bfbd5310b670caeb8](https://github.com/efroemling/ballistica/commit/63a835a1f4d4901481a4391bfbd5310b670caeb8)

## Screenshots
<details>
<summary>Windows application error report</summary>

![Screenshot 2021-08-09 215504](https://user-images.githubusercontent.com/42905362/128741569-78d6a1f4-30c0-4ec9-90ed-081c411a8eac.jpg)
</details>

<details>
<summary> Debug server console ouput</summary>

![2021-08-09](https://user-images.githubusercontent.com/42905362/128747000-27814a97-3dfc-434b-82c1-7a9052e2c86b.png)
</details>

## Scripts
<details>
<summary>portal.py:</summary>

```py
"""Defines Actor(s) related to snoball."""

from __future__ import annotations

from typing import TYPE_CHECKING
from dataclasses import dataclass

import random
import copy

import ba
from bastd.gameutils import SharedObjects

if TYPE_CHECKING:
    from typing import Any, List, Tuple, Sequence


@dataclass
class SpazTouchedMessage:
    """Message to the portal node when touches a Spaz."""
    portal_index: int = 0


@dataclass
class ObjectTouchedMessage:
    """Message to the portal node when touches an Object."""
    portal_index: int = 0


class PortalFactory:
    """A collection of objects used by Portal."""

    _STORENAME = ba.storagename()

    def __init__(self, activity: ba.GameActivity) -> None:
        """Instantiate a PortalFactory."""
        shared = SharedObjects.get()
        self.max_portals = 3
        self.current_num = 0
        self.active_pos: List[Tuple[float, float, float]] = []

        points = copy.copy(activity.map.ffa_spawn_points)
        points_2 = activity.map.powerup_spawn_points
        for point in points_2:
            points.append(point)
        pos: List[List[float]] = [[999, -999], [999, -999], [999, -999]]
        for point in points:
            for i in range(3):
                pos[i][0] = min(pos[i][0], point[i])
                pos[i][1] = max(pos[i][1], point[i])

        self.pos_bounds = pos

        self.portal_material_1 = ba.Material()
        self.portal_material_2 = ba.Material()

        self.portal_material_1.add_actions(
            conditions=(("we_are_younger_than", 500), "or",
                        ("they_are_younger_than", 100)),
            actions=("modify_node_collision", "collide", False))
        self.portal_material_1.add_actions(
            conditions=(("we_are_younger_than", 500), "or",
                        ("they_are_younger_than", 100)),
            actions=("modify_node_collision", "collide", False))

        self.portal_material_1.add_actions(
            conditions=("they_have_material", shared.player_material),
            actions=(("modify_part_collision", "collide", True),
                     ("modify_part_collision", "physical",
                      False), ("message", "our_node", "at_connect",
                               SpazTouchedMessage(0))))
        self.portal_material_2.add_actions(
            conditions=("they_have_material", shared.player_material),
            actions=(("modify_part_collision", "collide", True),
                     ("modify_part_collision", "physical",
                      False), ("message", "our_node", "at_connect",
                               SpazTouchedMessage(1))))

        self.portal_material_1.add_actions(
            conditions=(("they_have_material", shared.object_material), "and",
                        ("they_dont_have_material", shared.player_material)),
            actions=(("modify_part_collision", "collide",
                      True), ("modify_part_collision", "physical", False),
                     ("message", "our_node", "at_connect",
                      ObjectTouchedMessage(0))))
        self.portal_material_2.add_actions(
            conditions=(("they_have_material", shared.object_material), "and",
                        ("they_dont_have_material", shared.player_material)),
            actions=(("modify_part_collision", "collide",
                      True), ("modify_part_collision", "physical", False),
                     ("message", "our_node", "at_connect",
                      ObjectTouchedMessage(1))))

    @classmethod
    def get(cls) -> PortalFactory:
        """Return a shared ba.PowerupBoxFactory object, creating if needed."""
        activity = ba.getactivity()
        if activity is None or not isinstance(activity, ba.GameActivity):
            raise ba.ContextError("Must be called in a ba.GameActivity.")
        factory = activity.customdata.get(cls._STORENAME)
        if factory is None:
            factory = activity.customdata[cls._STORENAME] = PortalFactory(
                activity)
        assert isinstance(factory, PortalFactory)
        return factory


class Portal(ba.Actor):
    """A Portal class to teleport players, woooosh!"""

    def __init__(self,
                 position: Tuple[float, float, float] = None,
                 color: Sequence[float] = None,
                 owner: ba.Node = None,
                 r: float = 1.0):
        """Instantiate a portal."""
        super().__init__()
        factory = PortalFactory.get()

        self.radius = r
        self.position_1 = self.get_random_position(position)
        factory.active_pos.append(self.position_1)
        self.position_2 = self.get_random_position()
        factory.active_pos.append(self.position_2)
        self.positions = (self.position_1, self.position_2)

        self.recently_teleported: List[ba.Node] = []

        factory.current_num += 1

        clr: Sequence[float] = color or (random.random(), random.random(),
                                         random.random())

        self.node1 = ba.newnode("region",
                                delegate=self,
                                attrs={
                                    "position": self.position_1,
                                    "scale": (self.radius, self.radius,
                                              self.radius),
                                    "type": "sphere",
                                    "materials": [factory.portal_material_1]
                                })
        self.visual_radius_1 = ba.newnode("shield",
                                          attrs={
                                              "position": self.position_1,
                                              "color": clr,
                                              "radius": 0.1
                                          })
        ba.animate(self.visual_radius_1, "radius", {
            0: 0,
            0.5: self.radius * 2
        })
        ba.animate_array(self.node1, "scale", 3, {
            0: (0, 0, 0),
            0.5: (self.radius, self.radius, self.radius)
        })

        self.node2 = ba.newnode("region",
                                delegate=self,
                                attrs={
                                    "position": self.position_2,
                                    "scale": (self.radius, self.radius,
                                              self.radius),
                                    "type": "sphere",
                                    "materials": [factory.portal_material_2]
                                })
        self.visual_radius_2 = ba.newnode("shield",
                                          attrs={
                                              "position": self.position_2,
                                              "color": clr,
                                              "radius": 0.1
                                          })
        ba.animate(self.visual_radius_2, "radius", {
            0: 0,
            0.5: self.radius * 2
        })
        ba.animate_array(self.node2, "scale", 3, {
            0: (0, 0, 0),
            0.5: (self.radius, self.radius, self.radius)
        })

        # A simple check to see if the passed node is the source spaz taking
        # the portal powerup then send a stand message to go to our other end
        if owner is not None and position is not None and (owner.position
                                                           == position):
            owner.handlemessage(ba.StandMessage(position=self.node2.position))
            self.recently_teleported.append(owner)
            ba.timer(0.51, ba.Call(self.recently_teleported.remove, owner))

    def handlemessage(self, msg: Any) -> Any:
        if isinstance(msg, SpazTouchedMessage):
            node = ba.getcollision().opposingnode
            if node in self.recently_teleported:
                return None
            node.handlemessage(
                ba.StandMessage(position=self.positions[1 - msg.portal_index]))
            self.recently_teleported.append(node)
            ba.timer(0.01, ba.Call(self.recently_teleported.remove, node))
        elif isinstance(msg, ObjectTouchedMessage):
            node = ba.getcollision().opposingnode
            if node in self.recently_teleported:
                return None
            node.position = self.positions[1 - msg.portal_index]
            self.recently_teleported.append(node)
            ba.timer(0.01, ba.Call(self.recently_teleported.remove, node))
        elif isinstance(msg, ba.DieMessage):
            self.delete(msg.immediate)
        elif isinstance(msg, ba.OutOfBoundsMessage):
            self.handlemessage(ba.DieMessage(True, ba.DeathType.OUT_OF_BOUNDS))
        else:
            return super().handlemessage(msg)

        return None

    def delete(self, immediate: bool = False) -> None:
        """Deletes the portal."""
        if self.node1.exists() and self.node2.exists():
            factory = PortalFactory.get()
            factory.current_num -= 1
            self.node1.delete()
            self.node2.delete()
            factory.active_pos.remove(self.position_1)
            factory.active_pos.remove(self.position_2)
        if immediate:
            self.remove_visual()
        else:
            ba.animate(self.visual_radius_1, "radius", {
                0: self.radius * 2,
                0.5: 0
            })
            ba.animate(self.visual_radius_2, "radius", {
                0: self.radius * 2,
                0.5: 0
            })
            ba.timer(0.501, ba.WeakCall(self.remove_visual))

    def remove_visual(self) -> None:
        """Removes the portal visualizer shield."""
        if self.visual_radius_1.exists() and self.visual_radius_2.exists():
            self.visual_radius_1.delete()
            self.visual_radius_2.delete()

    def get_random_position(
            self,
            position: Sequence[float] = None) -> Tuple[float, float, float]:
        """Gets the random position on the map of current ba.GameActivity."""
        # The credit of this random position finder goes to Deva.
        # I only did some changes for not getting same positions again.
        factory = PortalFactory.get()
        r_u = random.uniform
        pos_b = factory.pos_bounds

        target: Sequence[float]
        if position is None:
            source: Sequence[float] = [
                r_u(pos_b[0][0] - 1.0, pos_b[0][1] + 1.0),
                pos_b[1][1] + r_u(0.1, 1.5),
                r_u(pos_b[2][0] - 1.0, pos_b[2][1] + 1.0)
            ]
            target = [source[0], source[1] - r_u(1.0, 1.3), source[2]]
        else:
            target = position

        count = 0
        # Checks if portal position is close to another one then finds new pos.
        # Limited the loop run times to 7 otherwise it could theoretically run
        # infinite number of times for not finding a valid position in bounds
        # when there are many many portals, just theoretical though.
        while any((ba.Vec3(target) - ba.Vec3(ac)).length() < self.radius * 2
                  for ac in factory.active_pos) and count < 7:
            count += 1
            source = [
                r_u(pos_b[0][0] - 1.0, pos_b[0][1] + 1.0),
                pos_b[1][1] + r_u(0.1, 1.5),
                r_u(pos_b[2][0] - 1.0, pos_b[2][1] + 1.0)
            ]
            target = [source[0], source[1] - r_u(1.0, 1.3), source[2]]

        return (target[0], target[1], target[2])
```
</details>

**Comment by drov-drov** [comment]
Perhaps the problem is WSL - it contains a lot of problems associated with building

**Comment by rahulraman0108** [comment]
I tried making the build using WSL and running with windows explorer GUI that didn't help too.
So I am not sure but as you mention building with WSL could be the cause of the issue.

I will try to build and run a debug headless build on an ubuntu 20.04 LTS server from CLI and then try joining from my device, will update here how that goes tomorrow.

**Comment by rahulraman0108** [comment]
So I built and then ran a debug headless build on ubuntu 20.04 and it crashes at the same time except that the error code is different and it seems to give some useful info about the crash.
![image](https://user-images.githubusercontent.com/42905362/129303346-a513ec2d-a320-43f4-a460-13378d7d746e.png)

**Comment by drov-drov** [comment]
I can assume that the problem is in `ballistica/dynamics /rigit_body.cc`

**Comment by rahulraman0108** [comment]
Hey there, I was travelling yesterday so I was not able to do that and update it here.
I will try to test on the headless build with portal.py removed and some other checks I would like to do in different parts of the script today or tomorrow when I get the time. I will update this issue after that.

**Comment by rahulraman0108** [comment]
Hello there,
I am sorry for the long delay, I got busy with school works and stuff.
So I actually got rid of the error-causing code in the portal.py.

It was actually the ba.animate_array call to animate the sphere regions that was causing the error.
I guess setting the region scale after it is created is not working (at least in debug builds), that is just my guess though.

Also since my issue is resolved now as I found out what the error actually was and removed the error from the script, I am closing this issue.
Although I am not sure that not being able to animate region scale is the intended behaviour. But it is never used in the original BombSquad scripts to animate region scale so I guess it is intended.
